source/common/regular-expressions.js

```diff
@@ -14,6 +14,190 @@
+   *
+   * @return  {RegExp}              The compiled Regular Expression
+   */
+  'getCitationRE': function () {
+    return RegExp(
+      /(\[(?:[^[\]]*@[^[\]]+)\])|(?<=\s|^)(@[\p{L}\d_][\p{L}\d_:.#$%&\-+?<>~/]*)/.source,
+      'gu')
+  },
+
+  /**
+   * Returns a regular expression that matches MarkDown inline code.
+   *
+   * @return  {RegExp}              The compiled Regular Expression
+   */
+  'getCodeRE': function () {
+    return RegExp(
+      /`.*?`/.source,
+      'i')
+  },
+
+  /**
+   * Returns a regular expression that matches MarkDown Code Blocks.
+   *
+   * @return  {RegExp}              The compiled Regular Expression
+   */
+  'getCodeBlockRE': function () {
+    return RegExp(
+      /^\s{0,3}(`{3,}|~{3,})/.source
+    )
+  },
+
+  /**
+   * Returns a regular expression that matches footnotes for project export.
+   *
+   * @param   {boolean}  multiline  Whether to match multiline
+   *
+   * @return  {RegExp}              The compiled Regular Expression
+   */
+  'getFnExportRE': function (multiline = false) {
+    let flag = (multiline) ? 'm' : ''
+    return RegExp(
+      /\[\^([\w]+?)\]/.source,
+      // Necessary flags + optional multiline flag
+      'g' + flag)
+  },
+
+  /**
+   * Returns a regular expression that matches MarkDown footnote references.
+   *
+   * @return  {RegExp}              The compiled Regular Expression
+   */
+  'getFnRE': function () {
+    return RegExp(
+      /(\[\^([\da-zA-Z_-]+)\][^:]|\[\^([\da-zA-Z_-]+)\]$)/.source,
+      'g')
+  },
+
+  /**
+   * Returns a regular expression that matches MarkDown footnote references.
+   *
+   * @return  {RegExp}              The compiled Regular Expression
+   */
+  'getFnReferenceRE': function () {
+    return RegExp(
+      /^\[\^.+\]:\s/.source
+    )
+  },
+
+  /**
+   * Returns a regular expression that matches MarkDown footnote references.
+   *
+   * @param   {boolean}  multiline  Whether to match multiline
+   *
+   * @return  {RegExp}              The compiled Regular Expression
+   */
+  'getFnRefRE': function (multiline = false) {
+    let flag = (multiline) ? 'm' : ''
+    return RegExp(
+      /^\[\^([\da-zA-Z_-]+)\]: (.+)/.source,
+      'g' + flag)
+  },
+
+  /**
+   * Returns a regular expression that matches MarkDown footnote references.
+   *
+   * @return  {RegExp}              The compiled Regular Expression
+   */
+  'getFootnoteRefRE': function () {
+    return RegExp(
+      /\[\^[^\]]+\]/.source,
+      'i')
+  },
+
+  /**
+   * Returns a regular expression that matches MarkDown Headings.
+   *
+   * @return  {RegExp}              The compiled Regular Expression
+   */
+  'getHeadRE': function () {
+    return RegExp(
+      /^(#{1,6}) (.*)/.source,
+      'g')
+  },
+
+  /**
+   * Returns a regular expression that matches MarkDown Headings.
+   *
+   * @return  {RegExp}              The compiled Regular Expression
+   */
+  'getHeadingRE': function () {
+    return RegExp(
+      /(#+)\s+/.source
```

Comment:
Don't we need to assure start-of-line with ^?

Reply:
I'll validate this against the current source tomorrow, but I did my best to make exact copies of the existing REs. Better to make changes once I'm sure I haven't broken anything!

Reply:
The current regex doesn't do that, so I won't change this for fear of unintended consequences, unless you'd prefer that I update it in this PR?

Reply:
No, no, if that breaks things, better not. I concur with you to better play it safe, especially with these big PRs!